### DISCUSSION TITLE
Springboot MVC Filter spans

### DISCUSSION THREAD

**Discussion by sarav-ks** [open-telemetry/opentelemetry-java-instrumentation#11988] [Q&A]
We are using 'Autoconfiguration' option to enable opentelemetry for our Springboot app (ie., Not using Java Agent instrumentation)

`    @Bean
    public OpenTelemetry openTelemetry() {
      return AutoConfiguredOpenTelemetrySdk.initialize().getOpenTelemetrySdk();
    }`

With the above configuration , i was expecting to see the Springboot spans like 'Filter' methods . For example i was expecting to see the span for 'OncePerRequestFilter' , which will be invoked for every API request . 

But i am not able to see these framework level spans in my API requests , though i can see the 'Controller' spans for every request.

Below is my maven dependencies for "Auto Configuration" 

`
        <dependency>
             <groupId>io.opentelemetry.instrumentation</groupId>
            <artifactId>opentelemetry-spring-boot-starter</artifactId>
        </dependency>     
         <dependency>
          <groupId>io.opentelemetry</groupId>
          <artifactId>opentelemetry-sdk-extension-autoconfigure</artifactId>
        </dependency>
        <dependency>
          <groupId>io.opentelemetry</groupId>
          <artifactId>opentelemetry-sdk-extension-autoconfigure-spi</artifactId>
        </dependency>  
`

Can you let me know if i have to explicitly add the dependencies as mentioned below to see the 'Filter' spans ?
https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/spring/spring-webmvc

**Comment by laurit** [comment]
> Can you let me know if i have to explicitly add the dependencies as mentioned below to see the 'Filter' spans ?
https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/spring/spring-webmvc

You don't need to add it, adding it won't get you the spans that you expect.
I think the problem here is that you have misunderstood what you will be getting. As far as I can tell what you are looking for is profiling  not tracing. Unfortunately otel java agent does not support profiling yet, but hopefully soon will.

**Reply by sarav-ks** [reply]
Thanks for the reply . Since we are planning to enable sampling of the traces , and inherently all the "slow" requests will be traced , i thought it will be great if "Spring Internal spans" can be captured as part of the "slow" trace , so we know where the time is spent.

Also , if there is no additional spans / metrics that we will getting generated using "Spring Otel Instrumentation" , what is the benefit of this versus "Java Agent Instrumentation"  other than the once mentioned at https://opentelemetry.io/docs/zero-code/java/spring-boot-starter/

**Reply by laurit** [reply]
> Also , if there is no additional spans / metrics that we will getting generated using "Spring Otel Instrumentation" , what is the benefit of this versus "Java Agent Instrumentation" other than the once mentioned at https://opentelemetry.io/docs/zero-code/java/spring-boot-starter/

You can use the spring boot starter when you don't wish to use the java agent. Java agents perform bytecode manipulation and can be a bit too much magic for some peoples taste. Spring boot starter offers automatic configuration similarly to the java agent.
If you need more spans you can always use manual instrumentation to create the spans that you need.

**Reply by sarav-ks** [reply]
Thanks for answering all my questions , appreciate your help .  But i am still not clear on the benefit of using 'Spring Boot' auto config ( other than the bytecode advantage) , would there be any additional metrics / spans from any of the modules of Spring . 

For example if we use 'Kafka'  , 'redis' modules of spring , would it not generate framework level 'spans' ?

**Reply by laurit** [reply]
As far as I know currently spring boot starter does not provide any metrics or spans that are not provided by java agent, though there are instrumentations in the java agent that are not present in the spring boot starter.

> For example if we use 'Kafka' , 'redis' modules of spring , would it not generate framework level 'spans' ?

Both the spring boot starter and java agent have instrumentations for kafka. Spring boot starter does not currently offer a redis instrumentation, java agent has instrumentations for lettuce, jedis, redission and rediscala, see https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/supported-libraries.md#libraries--frameworks
